### PR TITLE
Instrumental variable regression

### PR DESCRIPTION
**Pull request by nathanielolin**
Uses models fit with `AER::ivreg`. 

* Might be some cleanup to be done in `pred_se` around the `model.matrix` call. 
* Currently requires users to pass `data` since I didn't feel like refactoring the entire world to deal with what you get out of fitting the model with `x = TRUE` (a list of `model.frame` from each stage of the regression).
* Haven't tried weights, robust SEs, clustering...
* Needs all kinds of tests.

### CODE REVIEW

**Comment by nathanielolin** [comment]
@anniejw6 needs tests, and I don't think weights work yet, but is ready for a look.

File: R/ivreg.marg.R [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+      pred_margins = x$pred_margins,
+      se = x$se,
+      family = 'gaussian',
+      dof = dof,
+      cofint = c( (1 - cofint)/2, 1 - (1 - cofint)/2 )
+    )})
+
+}
+
+clean_glm_data <- function(mod, data, weights){
```
Comment (anniejw6) [inline R/ivreg.marg.R]: Delete this? Or maybe refactor it?

File: R/ivreg.marg.R [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+#' marg(mod, var_interest = 'actual', data = margex, type = 'levels')
+#' marg(mod, var_interest = 'actual', data = margex, type = 'effects')
+#'
+marg.ivreg <- function(mod, var_interest,
+                       type = 'levels',
+                       vcov_mat = NULL,
+                       dof = NULL,
+                       at = NULL, base_rn = 1,
+                       at_var_interest = NULL,
+                       data,
```
Comment (anniejw6) [inline R/ivreg.marg.R]: Can you move this up in the list? Typically, required params come before optional params.

Comment (anniejw6) [inline R/ivreg.marg.R]: Do you need a `if(is.null(data)) stop(...)`?

File: R/00_generics.R [hunk 3]
```diff
@@ -22,8 +22,8 @@
 #' If \code{NULL}, indicates all levels for a factor variable, defaults to \code{NULL}
 #' @param dof integer, the degrees of freedom used for the T statistic in an
 #' OLS model. Defaults to NULL in which case \code{mod$df.residual} is used.
-#' @param data data.frame that margins should run over, defaults to
-#' \code{mod$data}
+#' @param data data.frame that margins should run over. For glm method, defaults to
```
Comment (anniejw6) [inline R/00_generics.R]: Can you create an issue that we should investigate better ways of writing the documentation? It would be nice if `?marg` and `?marg.glm` and `?marg.ivreg` all led to different places?

File: R/ivreg.marg.R [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                       cofint = 0.95){
+  # Set params and Run checks ----
+
+  check_inputs(weights = weights, data = data, var_interest = var_interest,
+               at = at, cofint = cofint, base_rn = base_rn, type = type,
+               dof = dof, vcov_mat = vcov_mat)
+
+  # Checks specific to ivreg ----
+  # See if we're looking for continuous variables
+  if(type == 'effects' & is.numeric(data[[var_interest]]) &
```
Comment (anniejw6) [inline R/ivreg.marg.R]: Can you create an issue that we should find some ways of generalizing this?

File: R/ivreg.marg.R [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+     ! all(unique(data[[var_interest]]) %in% c(0, 1)) &
+     ! sprintf("as.character(%s)", var_interest) %in% names(data) &
+     ! sprintf("as.factor(%s)", var_interest) %in% names(data))
+    stop('We do not support effects for continuous variables at this time.')
+
+  # Check if no weights when model was built was weights
+  if(is.null(weights) & !is.null(mod$weights))
```
Comment (anniejw6) [inline R/ivreg.marg.R]: Can you create an issue that we should find some way of generalizing this?

File: R/ivreg.marg.R [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+#'
+#' v <- function(object, ...){
+#'   vcov(object) * object$df.residual / nobs(object)
+#' }
+#'
+#' summary(mod, vcov = v)
+#'
+#' # Get the level of the outcome variable at different values of `actual`
+#' # Note data is a *required* parameter here
+#' marg(mod, var_interest = 'actual', data = margex, type = 'levels')
```
Comment (anniejw6) [inline R/ivreg.marg.R]: In the `summary` function above, you use a custom `vcov` but not when you run `marg`. Why?

File: R/ivreg.marg.R [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    stop('We do not support effects for continuous variables at this time.')
+
+  # Check if no weights when model was built was weights
+  if(is.null(weights) & !is.null(mod$weights))
+    warning('The model was built with weights, but you have not ',
+            'provided weights. Your calculated margins may be odd. ',
+            'See Details.')
+
+  # Keep only complete variables ----
+  data <- clean_glm_data(mod, data, weights)
```
Comment (anniejw6) [inline R/ivreg.marg.R]: It seems like a lot of this is the same as `marg.glm`. Is there some better way of structuring `marg`?

File: R/ivreg.pred_se.R [hunk 8]
```diff
@@ -0,0 +1,52 @@
+# Main wrapper function to calculate margins and se
+pred_se.ivreg <- function(df_levels, model, type, base_rn, vcov_mat, weights){
+
```
Comment (anniejw6) [inline R/ivreg.pred_se.R]: Is the only thing that's different here from `glm` creating the model.frame?

File: R/ivreg.marg.R (hunk 6 above)
Comment (anniejw6) [inline R/ivreg.marg.R]: That seems a bit suboptimal to me, honestly. I wonder if we should just add a note / function allowing folks to use a custom variance-covariance matrix.

File: R/ivreg.marg.R (hunk 2 above)
Comment (nathanielolin) [inline R/ivreg.marg.R]: I tried to move it up the list for that reason, but the `cmd-check` yelled at me for violating S3 convention by having parameters in different orders... I wasn't sure what the best way to approach this is. 

I actually thought about setting the default to `NULL` but giving an error if `data` is `NULL` but that seemed... perverse.

File: R/ivreg.marg.R (hunk 7 above)
Comment (nathanielolin) [inline R/ivreg.marg.R]: I agree - I started feeling that way as I was modifying the function; I expected to need to tear up a lot more than was necessary. I do think there's probably a better way to structure marg.

But... should that be a different MR?

File: R/ivreg.marg.R (hunk 2 above)
Comment (anniejw6) [inline R/ivreg.marg.R]: why not just move up `data` in all of them?

File: R/ivreg.marg.R (hunk 7 above)
Comment (anniejw6) [inline R/ivreg.marg.R]: Yeah, different MR, but it'll probably make sense to work on that MR in parallel with this one...

File: R/ivreg.pred_se.R (hunk 8 above)
Comment (nathanielolin) [inline R/ivreg.pred_se.R]: Not quite. When calculating mean values it's not necessary to apply any link function (because we know it's gaussian), and the `deriv_func` function in the `calc_jacob` call is hard-coded since it's not in the model object (and doesn't vary).

File: R/ivreg.marg.R (hunk 6 above)
Comment (nathanielolin) [inline R/ivreg.marg.R]: Depends how strongly we feel about replicating Stata here. Personally, I'm happy to let the default results return `AER`'s standard errors, and include an example in the documentation with a `vcov_mat` that replicates Stata (like we do for clustering / robust SEs with `glm`s).

File: R/ivreg.marg.R (hunk 2 above)
Comment (nathanielolin) [inline R/ivreg.marg.R]: .................I can do that.

File: R/ivreg.marg.R (hunk 6 above)
Comment (anniejw6) [inline R/ivreg.marg.R]: Yeah, I think that's preferable. Maybe add it as a function...

File: R/ivreg.marg.R (hunk 2 above)
Comment (anniejw6) [inline R/ivreg.marg.R]: Was that dumb?

Comment (nathanielolin) [inline R/ivreg.marg.R]: No I just need to take a step back sometimes.

File: R/ivreg.marg.R (hunk 4 above)
Comment (nathanielolin) [inline R/ivreg.marg.R]: Done

File: R/ivreg.marg.R (hunk 5 above)
Comment (nathanielolin) [inline R/ivreg.marg.R]: Done

File: R/00_generics.R (hunk 3 above)
Comment (nathanielolin) [inline R/00_generics.R]: Done

**Comment by nathanielolin** [comment]
Closing this and moving ivreg implementation to a new branch that uses the new generic setup from #124
